cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramNotificationSubscriberService.java

```diff
@@ -263,27 +319,67 @@ private void handleProgramEvent(ProgramRunId programRunId, ProgramRunStatus prog
     RunRecordDetail recordedRunRecord;
     switch (programRunStatus) {
       case STARTING:
+        try {
+          RunRecordDetail runRecordDetail = appMetadataStore.getRun(programRunId);
+          if (runRecordDetail != null
+            && runRecordDetail.getStatus() != ProgramRunStatus.PENDING
+            && runRecordDetail.getStatus() != ProgramRunStatus.STARTING) {
+            //This is an invalid state transition happening. Valid state transitions are:
+            // PENDING => STARTING : normal state transition
+            // STARTING => STARTING : state transition after app-fabric restart
+            LOG.debug("Ignoring unexpected request to transition program run {} from {} state to program " +
+                        "STARTING state.", programRunId, runRecordDetail.getStatus());
+            return;
+          }
+        } catch (IllegalStateException ex) {
```

Comment:
when would this occur?

Reply:
This happens if there are multiple run records. AFAIK, this should never happen as it's been fixed, but regardless I put the catch here for safety.